multi_partner_learning.py

```diff
@@ -0,0 +1,532 @@
+            if self.learning_approach != 'seq':
+                clear_session()
+
+            # Split the train dataset in mini-batches
+            self.split_in_minibatches()
+
+            # Iterate over mini-batches and train
+            for minibatch_index in range(minibatch_count):
+
+                self.minibatch_index = minibatch_index
+
+                if self.learning_approach == 'fedavg':
+                    self.compute_collaborative_round_fedavg()
+
+                elif self.learning_approach == 'seq':
+                    self.compute_collaborative_round_sequential(sequentially_trained_model)
+
+                elif self.learning_approach == 'seqavg':
+                    self.compute_collaborative_round_seqavg()
+
+            # At the end of each epoch, evaluate the model for early stopping on a global validation set
+            # Even in the 'seq' approach we aggregate here...
+            # ... to minimize the dependence to the order of partners on last minibatch
+            self.prepare_aggregation_weights(self.local_score_list)
+            model_to_evaluate = self.build_aggregated_model(self.aggregate_model_weights())
+            model_evaluation = model_to_evaluate.evaluate(
+                x_val, y_val, batch_size=constants.DEFAULT_BATCH_SIZE, verbose=0,
+            )
+            current_val_loss = model_evaluation[0]
+            global_val_acc.append(model_evaluation[1])
+            global_val_loss.append(current_val_loss)
+            logger.info(f"   Model evaluation at the end of the epoch: "
+                        f"{['%.3f' % elem for elem in model_evaluation]}")
+
+            logger.info("      Checking if early stopping criteria are met:")
+            if is_early_stopping:
+                # Early stopping parameters
+                if (
+                        epoch_index >= constants.PATIENCE
+                        and current_val_loss > global_val_loss[-constants.PATIENCE]
+                ):
+                    logger.info("         -> Early stopping criteria are met, stopping here.")
+                    break
+                else:
+                    logger.info("         -> Early stopping criteria are not met, continuing with training.")
+
+        # After last epoch or if early stopping was triggered, evaluate model on the global testset
+        logger.info("### Evaluating model on test data:")
+        model_evaluation = model_to_evaluate.evaluate(
+            x_test, y_test, batch_size=constants.DEFAULT_BATCH_SIZE, verbose=0)
+        logger.info(f"   Model metrics names: {model_to_evaluate.metrics_names}")
+        logger.info(f"   Model metrics values: {['%.3f' % elem for elem in model_evaluation]}")
+        test_score = model_evaluation[1]  # 0 is for the loss
+
+        # Plot training history
+        if self.is_save_fig:
+
+            # Save data
+            history_data = {}
+            history_data["score_matrix"] = score_matrix
```

Comment:
in `update_iterative_results()` this is the attribute `self.score_matrix` which is updated.
I think we should remove the local variables `score_matrix`, `score_matrix_extented `and only use the attributes

Reply:
Yes, I am on it (I started working on basics that will be needed for the federated step-by-step increments, and it includes better score matrixes).